### PR TITLE
[AutoDiff] [SILVerifier] Verify `autodiff_function` instruction

### PR DESCRIPTION
**Pull request by rxwei**
* Verify the JVP type and VJP type in an `autodiff_function` instruction.
* Correct the associated function type calculator utility to produce closures with `@callee_guaranteed` callee convention.
* Add example code (/test/AutoDiff/autodiff_function_inst.sil) that demonstrates how we emit a thunk that forms a VJP function.

### CODE REVIEW

File: include/swift/AST/Types.h [hunk 1]
```diff
@@ -4077,8 +4077,8 @@ class SILFunctionType final : public TypeBase, public llvm::FoldingSetNode,
   CanSILFunctionType getWithDifferentiability(
       unsigned differentiationOrder, const SmallBitVector &parameterIndices);
 
-  /// Returns the differentiation function type associated functions with this
-  /// function type.
+  /// Returns the type of a differentiation fucntion that is associated with
```
Comment (marcrasi) [inline include/swift/AST/Types.h]: s/fucntion/function

File: lib/SIL/SILFunctionType.cpp [hunk 2]
```diff
@@ -290,7 +290,7 @@ CanSILFunctionType SILFunctionType::getAutoDiffAssociatedFunctionType(
               getAssociatedType(param.getType(), "CotangentVector")));
     auto pullbackType = SILFunctionType::get(getGenericSignature(),
                                              ExtInfo(), SILCoroutineKind::None,
-                                             ParameterConvention::Direct_Owned,
+                                         ParameterConvention::Direct_Guaranteed,
```
Comment (marcrasi) [inline lib/SIL/SILFunctionType.cpp]: this has become weirdly indented

File: lib/SIL/SILInstructions.cpp [hunk 3]
```diff
@@ -649,6 +649,8 @@ std::pair<SILValue, SILValue> AutoDiffFunctionInst::
 getAssociatedFunctionPair(unsigned differentiationOrder) const {
   assert(differentiationOrder > 0 &&
          differentiationOrder <= this->differentiationOrder);
+  assert(!getAssociatedFunctions().empty() &&
+         "Associated functions may not exist in raw SIL");
```
Comment (marcrasi) [inline lib/SIL/SILInstructions.cpp]: I don't understand this message. Does raw SIL have anything to do with this?

File: lib/SIL/SILVerifier.cpp [hunk 4]
```diff
@@ -1255,13 +1255,34 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
...
+    if (F.getModule().getStage() == SILStage::Canonical ||
+        adfi->hasAssociatedFunctions()) {
+      for (auto i : range(1, adfi->getDifferentiationOrder() + 1)) {
+        auto pair = adfi->getAssociatedFunctionPair(i);
+        auto jvpType = pair.first->getType().getAs<SILFunctionType>();
+        require(jvpType, "The JVP function must have a function type");
+        require(!jvpType->isDifferentiable(),
+                "The JVP function must not be @autodiff");
+        auto expectedJVPType = origTy->getAutoDiffAssociatedFunctionType(
+            adfi->getParameterIndices(), adfi->getDifferentiationOrder(),
```
Comment (marcrasi) [inline lib/SIL/SILVerifier.cpp]: adfi->getDifferentiationOrder() should be changed to i?

File: lib/SIL/SILVerifier.cpp [hunk 5]
```diff
@@ -1255,13 +1255,34 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
...
+            AutoDiffAssociatedFunctionKind::JVP, F.getModule());
+        require(expectedJVPType == jvpType, "Unexpected JVP function type");
+        auto vjpType = pair.second->getType().getAs<SILFunctionType>();
+        require(vjpType, "The VJP function must have a function type");
+        require(!vjpType->isDifferentiable(),
+                "The VJP function must not be @autodiff");
+        auto expectedVJPType = origTy->getAutoDiffAssociatedFunctionType(
+            adfi->getParameterIndices(), adfi->getDifferentiationOrder(),
```
Comment (marcrasi) [inline lib/SIL/SILVerifier.cpp]: adfi->getDifferentiationOrder() should be changed to i?

**Review by marcrasi — APPROVED** [review]

File: lib/SIL/SILFunctionType.cpp (hunk 2 above)
Comment (rxwei) [reply lib/SIL/SILFunctionType.cpp]: This is to fit within 80 columns. It's a pretty common, but weird, pattern within this code base.

File: lib/SIL/SILInstructions.cpp (hunk 3 above)
Comment (rxwei) [reply lib/SIL/SILInstructions.cpp]: When associated functions are not specified, the differentiation pass will fill them in. When the stage is raw, we cannot assume AD has already run.

Comment (rxwei) [reply lib/SIL/SILInstructions.cpp]: Clarified.

File: lib/SIL/SILFunctionType.cpp (hunk 2 above)
Comment (rxwei) [reply lib/SIL/SILFunctionType.cpp]: Re-formatted.

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux CPU

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux CPU

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux CPU

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux CPU

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux CPU

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux CPU

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux CPU

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux CPU

**Comment by rxwei** [comment]
@swift-ci please test tensorflow Linux

**Pull request merged**